### PR TITLE
Pp 447 patch email

### PR DESCRIPTION
**Pull request by mrlumbu**
## WHAT

_A brief description of the pull request:_
2 bits to this:
- A user may wish to add an email to a charge after the charge is created.
  This work adds a patch method to the charges/{chargeId} endpoint to allow this.
  At the moment the only patch operation allowed is "replace" and the only
  patchable field on the Charge entity is the email field.
- Refactor all Map request objects to be of type String, String, and remove a bunch of toString()'s
## HOW

_Steps to test or reproduce:_
Check tests cover all patch scenarios.
## WHO

_Who should review this pull request:_
- [x] Developers
- [] WebOps

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/model/ChargePatchRequest.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+import static uk.gov.pay.connector.model.PatchRequest.PatchField.*;
+
+public class ChargePatchRequest extends PatchRequest {
+    private static final List<String> patchableFields = Arrays.asList(ChargesApiResource.EMAIL_KEY);
+    private static final  List<String> validOps = Arrays.asList("replace");
```
Comment (imranbohoran) [inline src/main/java/uk/gov/pay/connector/model/ChargePatchRequest.java]: I feel these is best provided as a configuration or initialisation of the patch request object. If you have a PatchRequestBuilder and build it with the necessary patch fields and valid operations, we wouldn't need the `isValid(String)` and `isValidPath(String) overrides as well.
We could even get rid of the whole ChargePatchRequest class. What do you think?

File: src/main/java/uk/gov/pay/connector/resources/ChargesApiResource.java [hunk 2]
```diff
@@ -130,7 +129,7 @@ public Response getChargesJson(@PathParam(ACCOUNT_ID) Long accountId,
     @POST
     @Path(CHARGES_API_PATH)
     @Produces(APPLICATION_JSON)
-    public Response createNewCharge(@PathParam(ACCOUNT_ID) Long accountId, Map<String, Object> chargeRequest, @Context UriInfo uriInfo) {
+    public Response createNewCharge(@PathParam(ACCOUNT_ID) Long accountId, Map<String, String> chargeRequest, @Context UriInfo uriInfo) {
```
Comment (imranbohoran) [inline src/main/java/uk/gov/pay/connector/resources/ChargesApiResource.java]: 👍  properly typed Maps FTW

File: src/main/java/uk/gov/pay/connector/model/ChargePatchRequest.java (hunk 1 above)
Comment (mrlumbu) [inline src/main/java/uk/gov/pay/connector/model/ChargePatchRequest.java]: hmm ok, sounds good. Might need to check in with you about best way of doing that

**Comment by imranbohoran** [comment]
👍 For the builder. One last minor comment. There's an unused variable at line 40 of ChargeFrontendResource.

**Comment by imranbohoran** [comment]
👍

**Pull request merged**